src/com/amazon/corretto/crypto/provider/AesKeyWrapSpi.java

```diff
@@ -0,0 +1,227 @@
+        }
+        if (key != jceKey) {
+            if (!(key instanceof  SecretKey)) {
+                throw new InvalidKeyException("Need a SecretKey");
+            }
+            if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+                throw new InvalidKeyException("Need a raw format key");
+            }
+            if (!"AES".equalsIgnoreCase(key.getAlgorithm())) {
+                throw new InvalidKeyException("Expected an AES key");
+            }
+            if (keyBytes != null) {
+                Arrays.fill(keyBytes, (byte) 0);
+                keyBytes = null;
+            }
+            keyBytes = key.getEncoded();
+            if (keyBytes == null) {
+                throw new InvalidKeyException("Key doesn't support encoding");
+            }
+            jceKey = (SecretKey) key;
+        }
+        this.opmode = opmode;
+        // TODO [childw] do something with |random| here?
+    }
+
+    @Override
+    protected byte[] engineWrap(final Key key) throws IllegalBlockSizeException, InvalidKeyException {
+        if (opmode != Cipher.WRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] encoded = Utils.encodeForWrapping(provider, key);
+            final byte[] wrappedKey = new byte[4096];
+            int wrappedKeyLen = wrapKey(keyBytes, encoded, wrappedKey);
+            final byte[] wrappedKeyTruncated = new byte[wrappedKeyLen];
+            System.arraycopy(wrappedKey, 0, wrappedKeyTruncated, 0, wrappedKeyLen);
+            return wrappedKeyTruncated;
+        // TODO [childw] make this exception more specific, perhaps BadPaddingException?
+        } catch (final Exception ex) {
+            throw new InvalidKeyException("Wrapping failed", ex);
+        }
+    }
+
+    @Override
+    protected Key engineUnwrap(final byte[] wrappedKey, final String wrappedKeyAlgorithm, final int wrappedKeyType)
+            throws InvalidKeyException, NoSuchAlgorithmException {
+        if (opmode != Cipher.UNWRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] unwrappedKey = new byte[wrappedKey.length];
+            int unwrappedKeyLen = unwrapKey(keyBytes, wrappedKey, unwrappedKey);
+            final byte[] unwrappedKeyTruncated = new byte[unwrappedKeyLen];
+            System.arraycopy(unwrappedKey, 0, unwrappedKeyTruncated, 0, unwrappedKeyLen);
+            return Utils.buildUnwrappedKey(provider, unwrappedKeyTruncated, wrappedKeyAlgorithm, wrappedKeyType);
+        // TODO [childw] make this exception more specific, perhaps BadPaddingException?
+        } catch (final Exception ex) {
+            throw new InvalidKeyException("Unwrapping failed", ex);
+        }
+    }
+
+    @Override
+    protected byte[] engineUpdate(byte[] in, int inOffset, int inLen) {
```

Comment:
FYI, it looks like the [JCE supports ENCRYPT and DECRYPT](https://github.com/corretto/corretto-17/blob/4922f0805033d2f4a872add164f05320ef1592d3/src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java) as well as update/doFinal. We need to figure out our compatibility story. You might need to support those also. It should be relatively simple to do so (using AccessibleByteArrayOutputStream).

Reply:
yeah, i was hoping we could get away without buffering stuff across calls and managing that state/reset logic. if we need to support those Cipher modes as well, i guess we can't get around that. i'll start adding the buffering logic.